xla/service/gpu/runtime/nccl_collective_permute_thunk.cc

```diff
@@ -192,18 +225,30 @@ absl::Status RunCollectivePermute(
     TF_RETURN_IF_ERROR(nccl_api->GroupStart());
   }
 
-  // Send source buffer to target peer if needed.
-  if (target_id) {
-    TF_RETURN_IF_ERROR(nccl_api->Send(src_addr, buffer.element_type,
-                                      buffer.element_count, *target_id, comm,
-                                      &stream));
-  }
-
-  // Receive data from the source peer to the destination buffer.
-  if (source_id) {
-    TF_RETURN_IF_ERROR(nccl_api->Recv(dest_addr, buffer.element_type,
-                                      buffer.element_count, *source_id, comm,
-                                      &stream));
+  // If all peers are local, only get/send device pointer values and invoke
+  // memcpy.
+  if (use_memcpy) {
+    // If sending to another peer, get the pointer value of the src addr.
+    // Only change the pointer value when it's different from stored one.
+    if (source_id && recv_ptr_map[current_id].IsUnavailable()) {
+      VLOG(3) << "Using memcpy, initializing pointer: " << dest_addr.opaque()
+              << " current_id " << current_id;
+      recv_ptr_map[current_id].emplace(dest_addr.opaque());
```

Comment:
If somehow current_id will not be in the map, this will be a data race. If you use find and fail on missing id's then it should be safe.